packages/aws-cdk/lib/api/hotswap/step-functions.ts

```diff
@@ -0,0 +1,113 @@
+    machineName = undefined;
+  }
+
+  return new StepFunctionHotswapOperation({
+    logicalId: logicalId,
+    definition: stepDefinitionChange,
+    stateMachineName: machineName,
+  });
+}
+
+/**
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if the change is not for a `AWS::StepFunctions::StateMachine`,
+ * but doesn't prevent short-circuiting
+ * (like a change to CDKMetadata resource),
+ * `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if the change is to a AWS::StepFunctions::StateMachine,
+ * but not only to its definition property,
+ * or the definition string if the change is to a AWS::StepFunctions::StateMachine,
+ * and only affects its definition property.
+ */
+function isStepFunctionDefinitionOnlyChange(
+  change: cfn_diff.ResourceDifference, assetParamsWithEnv: { [key: string]: string },
+): string | ChangeHotswapImpact {
+  // if we see a different resource type, it will be caught by isNonHotswappableResourceChange()
+  // this also ignores Metadata changes
+  const newResourceType = change.newValue?.Type;
+  if (newResourceType !== 'AWS::StepFunctions::StateMachine') {
+    return ChangeHotswapImpact.IRRELEVANT;
+  }
+
+  if (change.oldValue?.Type == null) {
+    // can't short-circuit a brand new StateMachine
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  const propertyUpdates = change.propertyUpdates;
+
+  // ensure that only changes to the definition string result in a hotswap
+  for (const updatedPropName in propertyUpdates) {
+    const updatedProp = propertyUpdates[updatedPropName];
+
+    for (const newPropName in updatedProp.newValue) {
+      if (newPropName === 'Fn::Join') {
+        const joinString = updatedProp.newValue[newPropName];
+        const updatedDefinition = JSON.parse(joinString[1]);
+
+        return stringifyPotentialCfnExpression(JSON.stringify(updatedDefinition), assetParamsWithEnv);
+      }
+
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
+  }
+
+  return ChangeHotswapImpact.IRRELEVANT;
+}
+
+interface StepFunctionResource {
+  readonly logicalId: string;
+  readonly stateMachineName?: string;
+  readonly definition: string;
+}
+
+class StepFunctionHotswapOperation implements HotswapOperation {
+  constructor(private readonly stepFunctionResource: StepFunctionResource) {
+  }
+
+  public async apply(sdk: ISDK, stackResources: ListStackResources): Promise<any> {
+    let stateMachineName: string;
+    if (this.stepFunctionResource.stateMachineName) {
+      stateMachineName = this.stepFunctionResource.stateMachineName;
+    } else {
+      const stackResourceList = await stackResources.listStackResources();
+      const foundMachineName = stackResourceList
+        .find(resSummary => resSummary.LogicalResourceId === this.stepFunctionResource.logicalId)
+        ?.PhysicalResourceId;
+      if (!foundMachineName) {
+        // if we couldn't find the state machine in the current stack, we can't update it
+        return;
+      }
```

Comment:
This is copy&paste from the Lambda hotswapping. Can you extract this as a method on `ListStackResources` to get rid of this duplication?